File: internal/api/middleware/csrf.go
package middleware

import (
	"crypto/rand"
	"encoding/base64"
	"net/http"
	"os"
	"strings"

	"github.com/gin-gonic/gin"
)

func randomToken(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

// EnsureCSRFCookie sets a csrf_token cookie if it's missing.
//
// Uses SameSite=Strict to reduce cross-site cookie sending.
// HttpOnly must be false so browser JS can read it and send it back in a header.
func EnsureCSRFCookie() gin.HandlerFunc {
	return func(c *gin.Context) {
		if tok, err := c.Cookie("csrf_token"); err != nil || tok == "" {
			if tok, err := randomToken(32); err == nil {
				secure := os.Getenv("USE_HTTPS") == "true"
				http.SetCookie(c.Writer, &http.Cookie{
					Name:     "csrf_token",
					Value:    tok,
					Path:     "/",
					Secure:   secure,
					HttpOnly: false,
					SameSite: http.SameSiteStrictMode,
				})
			}
		}
		c.Next()
	}
}

// CSRFMiddleware enforces CSRF checks on unsafe methods for cookie-authenticated requests.
//
// - Skips safe methods (GET/HEAD/OPTIONS).
// - Skips requests using Authorization: Bearer (non-cookie API clients).
// - Enforces only when auth_token cookie is present (browser session).
//
// Validation uses the double-submit cookie pattern:
// cookie csrf_token must match X-CSRF-Token header OR _csrf form field.
func CSRFMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		m := c.Request.Method
		if m == http.MethodGet || m == http.MethodHead || m == http.MethodOptions {
			c.Next()
			return
		}

		if strings.HasPrefix(c.GetHeader("Authorization"), "Bearer ") {
			c.Next()
			return
		}

		// Only enforce when cookie auth is in play
		if _, err := c.Cookie("auth_token"); err != nil {
			c.Next()
			return
		}

		// Extra hardening: basic Origin check when present.
		if origin := c.GetHeader("Origin"); origin != "" {
			host := c.Request.Host
			if !strings.Contains(origin, host) {
				c.AbortWithStatusJSON(http.StatusForbidden, gin.H{"error": "csrf origin blocked"})
				return
			}
		}

		cookieTok, err := c.Cookie("csrf_token")
		if err != nil || cookieTok == "" {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{"error": "missing csrf cookie"})
			return
		}

		reqTok := c.GetHeader("X-CSRF-Token")
		if reqTok == "" {
			reqTok = c.PostForm("_csrf")
		}

		if reqTok == "" || reqTok != cookieTok {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{"error": "bad csrf token"})
			return
		}

		c.Next()
	}
}

File: src/server/middleware/csrf.ts
import { randomBytes } from 'node:crypto';
import type { NextFunction, Request, RequestHandler, Response } from 'express';
import { config } from '../config';

const SAFE_METHODS = new Set(['GET', 'HEAD', 'OPTIONS']);

/**
 * Issues the csrf_token cookie when missing.
 *
 * SameSite=Strict limits cross-site sending; HttpOnly must stay false so page
 * scripts can read the value back out and echo it in a header or form field.
 */
export function ensureCsrfCookie(): RequestHandler {
  return (req: Request, res: Response, next: NextFunction) => {
    if (!req.cookies?.['csrf_token']) {
      const token = randomBytes(32).toString('base64url');
      res.cookie('csrf_token', token, {
        path: '/',
        secure: config.useHttps,
        httpOnly: false,
        sameSite: 'strict',
      });
      // Make it visible to handlers within this same request.
      req.cookies = { ...req.cookies, csrf_token: token };
    }
    next();
  };
}

/**
 * Double-submit cookie check on unsafe methods for cookie-authenticated calls.
 *
 * - safe methods are skipped
 * - `Authorization: Bearer` clients are skipped (they are not cookie-driven)
 * - only enforced once an auth_token cookie is in play
 */
export function csrfMiddleware(): RequestHandler {
  return (req: Request, res: Response, next: NextFunction) => {
    if (SAFE_METHODS.has(req.method)) return next();
    if ((req.get('authorization') ?? '').startsWith('Bearer ')) return next();
    if (!req.cookies?.['auth_token']) return next();

    const origin = req.get('origin');
    if (origin && !origin.includes(req.get('host') ?? '')) {
      res.status(403).json({ error: 'csrf origin blocked' });
      return;
    }

    const cookieToken = req.cookies['csrf_token'];
    if (!cookieToken) {
      res.status(403).json({ error: 'missing csrf cookie' });
      return;
    }

    const requestToken = req.get('x-csrf-token') || req.body?.['_csrf'];
    if (!requestToken || requestToken !== cookieToken) {
      res.status(403).json({ error: 'bad csrf token' });
      return;
    }

    next();
  };
}
